beacon-chain/slasher/chunks.go

```diff
@@ -0,0 +1,553 @@
+			ctx, validatorIdx, minTarget,
+		)
+		if err != nil {
+			return nil, errors.Wrapf(
+				err, "could not get existing attestation record at target %d", minTarget,
+			)
+		}
+		if existingAttRecord != nil {
+			if sourceEpoch < existingAttRecord.IndexedAttestation.Data.Source.Epoch {
+				surroundingVotesTotal.Inc()
+				return &ethpb.AttesterSlashing{
+					Attestation_1: attestation.IndexedAttestation,
+					Attestation_2: existingAttRecord.IndexedAttestation,
+				}, nil
+			}
+		}
+	}
+	return nil, nil
+}
+
+// CheckSlashable takes in a validator index and an incoming attestation
+// and checks if the validator is slashable depending on the data
+// within the max span chunks slice. Recall that for an incoming attestation, B, and an
+// existing attestation, A:
+//
+//  B surrounds A if and only if B.target < max_spans[B.source]
+//
+// That is, this condition is sufficient to check if an incoming attestation
+// is surrounded by a previous one. We also check if we indeed have an existing
+// attestation record in the database if the condition holds true in order
+// to be confident of a slashable offense.
+func (m *MaxSpanChunksSlice) CheckSlashable(
+	ctx context.Context,
+	slasherDB db.SlasherDatabase,
+	validatorIdx types.ValidatorIndex,
+	attestation *slashertypes.IndexedAttestationWrapper,
+) (*ethpb.AttesterSlashing, error) {
+	sourceEpoch := attestation.IndexedAttestation.Data.Source.Epoch
+	targetEpoch := attestation.IndexedAttestation.Data.Target.Epoch
+	maxTarget, err := chunkDataAtEpoch(m.params, m.data, validatorIdx, sourceEpoch)
+	if err != nil {
+		return nil, errors.Wrapf(
+			err, "could not get max target for validator %d at epoch %d", validatorIdx, sourceEpoch,
+		)
+	}
+	if targetEpoch < maxTarget {
+		existingAttRecord, err := slasherDB.AttestationRecordForValidator(
+			ctx, validatorIdx, maxTarget,
+		)
+		if err != nil {
+			return nil, errors.Wrapf(
+				err, "could not get existing attestation record at target %d", maxTarget,
+			)
+		}
+		if existingAttRecord != nil {
+			if existingAttRecord.IndexedAttestation.Data.Source.Epoch < sourceEpoch {
+				surroundedVotesTotal.Inc()
+				return &ethpb.AttesterSlashing{
+					Attestation_1: existingAttRecord.IndexedAttestation,
+					Attestation_2: attestation.IndexedAttestation,
+				}, nil
+			}
+		}
+	}
+	return nil, nil
+}
+
+// Update a min span chunk for a validator index starting at the current epoch, e_c, then updating
+// down to e_c - H where H is the historyLength we keep for each span. This historyLength
+// corresponds to the weak subjectivity period of eth2, see:
+// https://github.com/ethereum/eth2.0-specs/blob/dev/specs/phase0/weak-subjectivity.md.
+// This means our updates are done in a sliding window manner. For example, if the current epoch
+// is 20 and the historyLength is 12, then we will update every value for the validator's min span
+// from epoch 20 down to epoch 8 (since 20 - 12 = 8).
```

Comment:
```suggestion
// from epoch 20 down to epoch 9.
```
This is because we also count epoch 20. In code we have:
```
minEpoch = args.currentEpoch - (m.params.historyLength - 1)
```